frontend/app/.server/routes/security/validate-auth-session.ts

```diff
@@ -0,0 +1,45 @@
+import type { ActionFunctionArgs, LoaderFunctionArgs } from '@remix-run/node';
+import { redirectDocument } from '@remix-run/node';
+
+import type { PickDeep } from 'type-fest';
+
+import { TYPES } from '~/.server/constants';
+import { RaoidcSessionInvalidException } from '~/.server/web/exceptions';
+
+/**
+ * Validates the RAOIDC session for the incoming request.
+ *
+ * This function ensures that the RAOIDC session associated with the request is valid.
+ * It retrieves the RAOIDC session validator from the application context and validates the session.
+ *
+ * If the session is invalid, it redirects the user to the login page, preserving the
+ * current request's URL as a `returnto` query parameter for post-login redirection.
+ * Any other errors encountered during validation are re-thrown.
+ *
+ * **Usage in Remix Loader or Action Functions:**
+ *
+ * ```typescript
+ * import { validateAuthSession } from '~/.server/routes/security';
+ *
+ * export const loader = async ({ request, context }: LoaderFunctionArgs) => {
+ *   // Validate the RAOIDC session first
+ *   await validateAuthSession({ context, request });
+ *
+ *   // ... continue processing the loader logic
+ *   return { success: true };
+ * };
+ * ```
+ */
+export async function validateAuthSession({ context, request }: PickDeep<ActionFunctionArgs | LoaderFunctionArgs, 'context.appContainer' | 'request'>): Promise<void> {
```

Comment:
I don't really like this.. I see no benefit to moving this code outside of the route that uses it.

If you look at the route now, you see `await validateAuthSession({ context: { appContainer }, request })` with no indication that a redirect might be thrown. It makes the code more difficult to read.

And I could argue that the call to `context.appContainer` is dependency lookup, not dependency injection. Why should this function know *anything* about a remix context or an inversify context?

Reply:
Having seen the number of times this function is called, I understand that it exists to keep things DRY.

I guess since we don't have the benefit of using middleware in Remix (yet), this is the best you can do.

But I still don't think this function should know anything about contexts or app containers. The validator should be injected, not the container.

Reply:
I've added `SecurityHandler` from `~/.server/routes/security`. It can be retreived from `appContainer` and used like this:

```ts
const securityHandler = appContainer.get(TYPES.routes.security.SecurityHandler);
await securityHandler.validateAuthSession(request);
```